pkg/terraform/store.go

```diff
@@ -122,6 +141,9 @@ func (ws *WorkspaceStore) Workspace(ctx context.Context, c resource.SecretClient
 	}
 	l := ws.logger.WithValues("workspace", dir)
 	ws.mu.Lock()
+	if err := ws.startSharedServer(); err != nil {
```

Comment:
Another option could be to decouple the lifecycle of the provider server from the operations, i.e. maybe a struct different than `WorkspaceStore` and called directly in `cmd/main.go` of the providers and can restart the server if automatically if it crashes. The advantage would be that these parts would stay independent of the deployment model of the provider. Do you think that would be feasible?

Reply:
I think that should be feasible, if I understand you correctly, we could move anything related we do for the shared gRPC server to its own struct. But in order to keep things simpler I did not want to have a separate lifecycle management for the shared server. Ideally, I would have preferred a separate K8s deployment and a corresponding TCP service, which we can scale independently from the Crossplane provider. Or at least a sidecar container for which we could have separate liveness checks to have it managed by kubelet. I think coupling the lifecycle of the shared server with the workspace store strikes a good balance between simplicity & robustness and it's very similar to Terraform's production model, where the Terraform CLI automatically forks it before each call. With the proposed implementation, if we have some extra configuration for the workspacestore (namely the path of the binary plugin and the args to be passed to the native plugin process), with no separate logic to be implemented we have a robust, auto-starting server available. Do you think we would need a separate lifecycle for the shared server than the corresponding workspacestore? 

If you also agree with the initial demo deployment of this feature (where we had a separate K8s deployment for the shared server behind a K8s service), then I would propose to keep the lifecycle bound to workspacestore as of now until we have the necessary supporting features in the package manager to implement that. I like the simplicity we have now.

Reply:
I have a slight preference for decoupling but I can see your point. Maybe then we can add an interface like `TerraformProviderRunner` with a `StartProvider` function? It would be called by `WorkspaceStore` just like here but it'd be satisfied by a `NopRunner` by default for example. Users can toggle what runner they want when they initialize the `WorkspaceStore` and the runners would be configured independently.

Reply:
Done. I've added a `terraform. NativeProviderRunner` interface and the corresponding `providerRunner` member to the `terraform.WorkspaceStore` struct. By default a `terraoform. NoOpProviderRunner` is used.